### ISSUE TITLE
Getting NANs with arbitrary attn_bias mask with xformers==0.0.18 

### ISSUE THREAD

**Issue by toothacher17** [facebookresearch/xformers#722]
# 🐛 Bug

I am trying to use xformers to replace my native pytorch MHA implementation, sth like:

```
scale = 1 / query.shape[-1] ** 0.5
query = query * scale
attn = query @ key.transpose(-2, -1)
if attn_bias is not None:
    attn = attn + attn_bias
attn = attn.softmax(-1)
attn = F.dropout(attn, p)
return attn @ value

```

After switching to xformers, I am usiing xops.memory_efficient_attention(q, k ,v, attn_bias).

This works fine when I am using a lower triangular mask, either by passing in a LowerTriangularMask() or passing a torch.tensor with the same shape that built by my own.

However, when I am switching to use a arbitrary mask (supposing in pretraining stage, you opened `reset_position_ids` and `reset_attention_mask` flags, so you'll get a new start inside one sequence), I am getting NANs during evaluating (no grad forward) or training (with grad). Based on the log, the program is using the CUTLASS op.

Based on my observations, xformers saves **10-15% GPU memory** and improves overall **TFLOPs by 10-15%**, so I really want to use it to replace with my native pytorch implementations. Could you help on this issue? 

## Environment

some depenedencies
trition==2.0
xformers==0.0.18
pytorch==2.0

**Comment by toothacher17** [comment]
So my question is: is arbitrary attention mask supported by xformers 0.0.18 yet?

**Comment by danthe3rd** [comment]
Hi @toothacher17 
This should be supported by xFormers, but the behavior you report is definitively a bug. Unless a line in your mask is entirely masked-out, it shouldn't give you NaNs.
Do you have an independant minimum repro example so I can try it?

**Comment by toothacher17** [comment]
hi, @danthe3rd 

Thanks a lot for the reply. 

Yes, I took another look at the BlockDiagonalCausalMask at flash_attention and Megatron-LM repo a few days ago. Basically it meets my need of pretraining with reset-position-id and reset-attention-mask, as resetting positions changes a lower-triangular matrix to a block diagonal matrix. To use this mask, like Megatron-LM, I'll need to change the shape, and merge all sentences in a batch into a single sentence, calculate the cumulative sequence length and pass them in to the ops.

**Comment by toothacher17** [comment]
Btw, I tested with cutlass based operators as it is the only class that supports customized attention bias. And the NANs is generated by the cutlass operator...

**Comment by toothacher17** [comment]
Thanks for getting back to me. I am using sth like ROPE, but that's before it enters into attention, so as long as the cumulative sequence length for qkv is calculated correctly, the block diagonal causal matrix should be fine.

How soon do you think you can release the fix? If it is really coming soon, I'll wait for the fix before changing my code for the block diagonal causal matrix

**Comment by danthe3rd** [comment]
It should be fixed as of https://github.com/facebookresearch/xformers/commit/540fcbfdaf8230fd37644345a310a36af86c9d6b, and will be included in the next release (**0.0.19**). In the meantime, you can also use a development build `>=0.0.19.dev516`

**Comment by toothacher17** [comment]
Thanks, @danthe3rd . We managed to change to use flash_atten_unppad_func and get correct expected loss. Maybe after the 0.0.19 is released, we might change back to use xformers and see if cutlass is faster than flash_attn implementation.

Thanks for the quick fix! Cheers!

**Comment by nofreewill42** [comment]
Is an arbitrary mask with i.e. an off diagonal rectangular shaped attention mask being memory efficient if my understanding is correct, but it computes also the grey area as well so compute efficiency is lacking?
I'm awaiting nervously for this feature :P because of cross attention to audio from a transcript text efficiently like in this image:
![image](https://github.com/facebookresearch/xformers/assets/14865017/356c8d0a-6f83-4401-ab04-bd751206a2d3)
